ruma-common/src/push.rs

```diff
@@ -94,27 +94,62 @@ impl Ruleset {
     }
 }
 
+/// Iterator type for `Ruleset`
+#[derive(Debug)]
+pub struct RulesetIter {
+    content: BTreeSetIter<ContentPushRule>,
+    override_: BTreeSetIter<OverridePushRule>,
+    room: BTreeSetIter<RoomPushRule>,
+    sender: BTreeSetIter<SenderPushRule>,
+    underride: BTreeSetIter<UnderridePushRule>,
+}
+
+impl Iterator for RulesetIter {
+    type Item = AnyPushRule;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.content
+            .next()
+            .map(|x| x.0.into())
+            .or_else(|| self.override_.next().map(|x| x.0.into()))
+            .or_else(|| self.room.next().map(|x| x.0.into()))
+            .or_else(|| self.sender.next().map(|x| x.0.into()))
+            .or_else(|| self.underride.next().map(|x| x.0.into()))
+    }
+}
+
 impl IntoIterator for Ruleset {
     type Item = AnyPushRule;
-    type IntoIter = std::vec::IntoIter<Self::Item>;
+    type IntoIter = RulesetIter;
 
     fn into_iter(self) -> Self::IntoIter {
-        self.content.into_iter().map(|x| x.0.into())
-            .chain(self.override_.into_iter().map(|x| x.0.into()))
-            .chain(self.room.into_iter().map(|x| x.0.into()))
-            .chain(self.sender.into_iter().map(|x| x.0.into()))
-            .chain(self.underride.into_iter().map(|x| x.0.into()))
-            .collect::<Vec<_>>().into_iter()
+        RulesetIter {
+            content: self.content.into_iter(),
+            override_: self.override_.into_iter(),
+            room: self.room.into_iter(),
+            sender: self.sender.into_iter(),
+            underride: self.underride.into_iter(),
+        }
     }
 }
 
 /// A trait for types that can be added in a Ruleset
-pub trait RulesetMember {
+pub trait RulesetMember: private::Sealed {
     /// Adds a value in the correct field of a Ruleset.
     fn add_to(self, ruleset: &mut Ruleset) -> bool;
```

Comment:
Can you also add `#[doc(hidden)]` to this function, since users aren't meant to call it directly?